http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala

```diff
@@ -527,7 +524,187 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
       }
     } yield ()
 
-  def createGalaxyPostgresDisk(project: GoogleProject, zone: ZoneName, namespaceName: NamespaceName): Disk =
+  override def stopAndPollApp(params: StopAppParams)(implicit ev: Ask[F, AppContext]): F[Unit] =
+    for {
+      ctx <- ev.ask
+
+      dbAppOpt <- KubernetesServiceDbQueries.getFullAppByName(params.googleProject, params.appId).transaction
+      dbApp <- F.fromOption(dbAppOpt, AppNotFoundException(params.googleProject, params.appName, ctx.traceId))
+      dbNodepool = dbApp.nodepool
+      dbCluster = dbApp.cluster
+      nodepoolId = NodepoolId(dbCluster.getGkeClusterId, dbNodepool.nodepoolName)
+
+      _ <- logger.info(
+        s"Stopping app ${dbApp.app.appName.value} in cluster ${dbCluster.getGkeClusterId.toString} | trace id: ${ctx.traceId}"
+      )
+
+      _ <- nodepoolQuery.updateStatus(dbNodepool.id, NodepoolStatus.Provisioning).transaction
+
+      // If autoscaling is enabled, disable it first
+      _ <- if (dbNodepool.autoscalingEnabled) {
+        nodepoolLock.withKeyLock(dbCluster.getGkeClusterId) {
```

Comment:
I think nodepoolLock can be `Ref[IO, Map[gkeClusterId, Semaphore[IO]]]`, and this can be 
```
permitOpt <- nodepoolLock.get(_.get(gkeClusterId))
permit <- permitOpt.fold {
        Semaphore[IO](1L).flatMap(s => nodepoolLock.modify(mp => (mp + (gkeClusterId -> s), s))) //make a new lock if not exist already
      }(p => IO.pure(p))
lock = Resource.make[IO, Unit](permit.acquire)(_ => permit.release)
_ <- lock.use {_ => ???}
```

This way we don't need the custom classes and mvar etc

Reply:
I think I tried `Ref[IO, Map[..]]` and had some issues.. but your code seems correct. In any case I think we'd want some helper code to make the lock easy to use.

Another reason I kinda like using guava cache is it gives us lock timeouts, to prevent potentially rogue processes from holding locks forever.

Reply:
ah yea, that makes sense....